File: CompilerFramework/ParserFramework.cs
//    CompilerFramework
//    Copyright(C) 2019  刘迅承

//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.

//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
//    GNU General Public License for more details.

//    You should have received a copy of the GNU General Public License
//    along with this program.If not, see<https://www.gnu.org/licenses/>.
using CompilerFramework.Lexer;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace CompilerFramework.Parser
{
    /// <summary>
    /// The very base framework of parsing, provide very base methods for all kind of parsing.
    /// </summary>
    class ParserFramework
    {
        /// <summary>
        /// For the async multi-thread parsing order chacking.
        /// </summary>
        protected ParseDoor door = new ParseDoor();
        /// <summary>
        /// For the parsing order chacking.
        /// </summary>
        protected long index = 0;
        /// <summary>
        /// stack for PDA
        /// </summary>
        protected Stack<object> stack;
        /// <summary>
        /// collection of productions
        /// </summary>
        public List<Production> Productions { get; private set; }
        /// <summary>
        /// Semant Delegate Table for string convert to delegate.
        /// </summary>
        public Dictionary<string, SemantDelegate> SemantDelegateTable = new Dictionary<string, SemantDelegate>();
        /// <summary>
        /// parsing entrance methods
        /// </summary>
        /// <param name="parseUnit">parse unit</param>
        protected virtual void Parse(ParseUnit parseUnit) { }
        #region receive result of lexing
        /// <summary>
        /// Receive LexerResult and Parse synchronously, order of Index will be check.
        /// </summary>
        /// <param name="lexerResult">The Result of Lexer</param>
        public void ParseLexUnit(LexerResult lexerResult)
        {
            //check the order
            if (lexerResult.Index != index) throw new ParseIndexException(index, lexerResult.Index,
                "Index Error: Index should be " + index + " instead of " + lexerResult.Index);
            //transfer to parse unit
            ParseUnit parseUnit = new ParseUnit(lexerResult.Name, null, lexerResult.Position, lexerResult.Value, null);
            //to parse it
            index++;// move to next for sync
            door.GetIn();// move to next for async, but no use in here.
            Parse(parseUnit);
        }
        /// <summary>
        /// Receive LexerResult and Parse asynchronously, order of Index will be checked.
        /// </summary>
        /// <param name="lexerResult">The Result of Lexer</param>
        public async Task ParseLexUnitAsync(LexerResult lexerResult)
        {
            //check the order
            if (lexerResult.Index != index) throw new ParseIndexException(index, lexerResult.Index,
                "Index Error: Index should be " + index + " instead of " + lexerResult.Index);
            //transfer to parse unit
            ParseUnit parseUnit = new ParseUnit(lexerResult.Name, null, lexerResult.Position, lexerResult.Value, null);
            //to parse it
            index++;// move to next for sync
            await JoinParse(parseUnit, lexerResult.Index);
        }
        protected Task JoinParse(ParseUnit parseUnit, long index)
        {
            Task task = new Task(new Action<object>(x => Parse((ParseUnit)x)), parseUnit);
            while (door.Index < index) { }
            if (door.Index == index)
            {
                lock (door)
                {
                    door.GetIn();// move to next for async
                    task.RunSynchronously();// run it exclusively
                }
            }
            else
            {
                throw new AsyncIndexException(index, door.Index,
                "Async Index Error: Async Index should be " + index + " instead of " + door.Index + ". Check your code please.");
            }
            return task;
        }
        #endregion
        #region receive productions
        /// <summary>
        /// the most direct way to add production
        /// </summary>
        /// <param name="production">production</param>
        public void AddProduction(Production production) => Productions.Add(production);
        /// <summary>
        /// part of indirect way of add production, add the SemantDelegate to table.
        /// </summary>
        /// <param name="name">name of delegate method</param>
        /// <param name="semantDelegate">the delegate method</param>
        /// <exception cref="ProductSentenceException">if ProductSentence is illeagal</exception>
        public void AddSemantDelegateTable(string name, SemantDelegate semantDelegate) => SemantDelegateTable.Add(name, semantDelegate);
        /// <summary>
        /// indirect way of add production by string, format string as:
        /// Pre ->|:|| Suf0 Suf1 ... [@SemantDelegateName[$attr]]
        /// like this:
        /// E -> F G
        /// or like this:
        /// E -> E add E @SemantAdd
        /// or like this:
        /// E | E opt E @SementOpt
        /// or like this:
        /// E : Delimiter E Delimiter @Semant
        /// or with attr in string (attr will be string type), just add $attr directly behand @SemantDelegateName
        /// E -> E F @Semant$deal
        /// </summary>
        /// <param name="productSentence">formated string</param>
        /// <exception cref="ProductSentenceException">if productSentence is illeagal</exception>
        public void AddProductionByString(string productSentence)
        {
            string[] vs = productSentence.Split(" ");
            if (vs[1] != "->" && vs[1] != "|" && vs[1] != ":") throw new ProductSentenceException(productSentence,
                    "ProductSentence Error: sentence \"" + productSentence + "\" is illeagal.");
            int i = vs.Length - 1;
            SemantDelegate semantDelegate = null;
            object attr = null;
            if (vs[i].StartsWith('@'))
            {
                string[] s = vs[i].TrimStart('@').Split("$", 2);
                semantDelegate = SemantDelegateTable[s[0]];
                if (s[1].Length > 0) attr = s[1];
                i--;
            }
            string[] sufs = new string[i];
            for (int j = 2; j <= i; j++){
                sufs[j - 2] = vs[j];
            }
            Production production = new Production(vs[0], sufs, semantDelegate, attr);
            AddProduction(production);
        }
        /// <summary>
        /// indirect way of add productions by strings, format string as:
        /// Pre ->|:|| Suf0 Suf1 ... [@SemantDelegateName[$attr]]
        /// like this:
        /// E -> F G
        /// or like this:
        /// E -> E add E @SemantAdd
        /// or like this:
        /// E | E opt E @SementOpt
        /// or like this:
        /// E : Delimiter E Delimiter @Semant
        /// or with attr in string (attr will be string type), just add $attr directly behand @SemantDelegateName
        /// E -> E F @Semant$deal
        /// </summary>
        /// <param name="productSentences">formated strings</param>
        public void AddProductionsByStrings(params string[] productSentences) { foreach (string s in productSentences) AddProductionByString(s); }
        #endregion
    }
    /// <summary>
    /// Exception produeced while ProductSentence is illeagal.
    /// </summary>
    public class ProductSentenceException : Exception
    {
        /// <summary>
        /// Error Sentence
        /// </summary>
        public string Sentence { get; }
        /// <summary>
        /// Construct menthod
        /// </summary>
        /// <param name="sentence">error sentence</param>
        /// <param name="message">error message</param>
        public ProductSentenceException(string sentence, string message) : base(message)
        {
            Sentence = sentence;
        }
    }
    /// <summary>
    /// Semant delegate and return a value for next step.
    /// </summary>
    /// <param name="parseUnits"></param>
    /// <returns></returns>
    public delegate object SemantDelegate(ParseUnit[] parseUnits);
    /// <summary>
    /// Unit of Parse, such as phrase.
    /// </summary>
    public struct ParseUnit
    {
        /// <summary>
        /// Unit of Parse, such as phrase.
        /// </summary>
        /// <param name="name">name of this, such as "class"</param>
        /// <param name="parseUnits">the unit contained, as son nodes</param>
        /// <param name="value">reserved place for advanced usage</param>
        /// <param name="property">reserved place for advanced usage</param>
        /// <param name="position">position of unit</param>
        internal ParseUnit(string name, ParseUnit[] parseUnits, Position position, object value, object property)
        {
            Name = name;
            ParseUnits = parseUnits;
            Value = value;
            Position = position;
            Property = property;
        }
        /// <summary>
        /// name of this
        /// </summary>
        public string Name { get; }
        /// <summary>
        /// the unit contained, as son nodes
        /// </summary>
        public ParseUnit[] ParseUnits { get; }
        /// <summary>
        /// reserved place for advanced usage
        /// </summary>
        public object Value { get; }
        /// <summary>
        /// position of unit
        /// </summary>
        public Position Position { get; }
        /// <summary>
        /// reserved place for advanced usage
        /// </summary>
        public object Property { get; }
    }
    /// <summary>
    /// Exception when index is not match
    /// </summary>
    public class ParseIndexException : Exception
    {
        /// <summary>
        /// Exception when index is not match
        /// </summary>
        /// <param name="corIndex">index should be</param>
        /// <param name="errIndex">current index</param>
        /// <param name="message">message of exception</param>
        public ParseIndexException(long corIndex, long errIndex, string message) : base(message)
        {
            CorIndex = corIndex;
            ErrIndex = errIndex;
        }
        /// <summary>
        /// index should be
        /// </summary>
        public long CorIndex { get; }
        /// <summary>
        /// current index
        /// </summary>
        public long ErrIndex { get; }
    }
    /// <summary>
    /// For async multi-thread operation to lock it
    /// </summary>
    internal class ParseDoor
    {
        /// <summary>
        /// which index should to lock next
        /// </summary>
        internal long Index { get; private set; } = 0;
        /// <summary>
        /// get in door and lock this object, lock operation should be done by you outside.
        /// </summary>
        internal void GetIn() => Index++;
    }
    /// <summary>
    /// Exception when async index is bigger than current index, generally this will not happen if no coding error.
    /// </summary>
    public class AsyncIndexException : Exception
    {
        /// <summary>
        /// Exception when async index is bigger than current index, generally this will not happen if no coding error.
        /// </summary>
        /// <param name="curIndex">current index</param>
        /// <param name="errIndex">error async index</param>
        /// <param name="message">message of exception</param>
        public AsyncIndexException(long curIndex, long errIndex, string message) : base(message)
        {
            CurIndex = curIndex;
            ErrIndex = errIndex;
        }
        /// <summary>
        /// current index
        /// </summary>
        public long CurIndex { get; }
        /// <summary>
        /// error async index
        /// </summary>
        public long ErrIndex { get; }
    }
    /// <summary>
    /// the production of grammar
    /// </summary>
    public struct Production
    {
        /// <summary>
        /// antecedent
        /// </summary>
        public string Pre { get; }
        /// <summary>
        /// seccedents
        /// </summary>
        public string[] Sufs { get; }
        /// <summary>
        /// delegate for semanting
        /// </summary>
        public SemantDelegate SemantDelegate { get; }
        /// <summary>
        /// attr for advansing usage
        /// </summary>
        public object Attr { get; }
        /// <summary>
        /// the production of grammar
        /// </summary>
        /// <param name="pre">antecedent</param>
        /// <param name="sufs">seccedents</param>
        /// <param name="semantDelegate">delegate for semanting</param>
        /// <param name="attr">attr for advansing usage</param>
        public Production(string pre, string[] sufs, SemantDelegate semantDelegate, object attr)
        {
            Pre = pre;
            Sufs = sufs;
            SemantDelegate = semantDelegate;
            Attr = attr;
        }
    }
}

File: PythonProject/parser_framework.py
#    CompilerFramework Python Version - LexerFramework
#    Copyright(C) 2023  刘迅承

#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.

#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
#    GNU General Public License for more details.

#    You should have received a copy of the GNU General Public License
#    along with this program.If not, see<https://www.gnu.org/licenses/>.

from typing import Callable, Any, List, Tuple, Dict
import time
from queue import LifoQueue

from lexer_framework import LexerResult


class ParseUnit:
    """
    Unit of Parse, such as phrase.

    Parameters
    ----------
    self.name : str
        name of this, such as "class"
    self.parse_units : :obj:`list` of :obj:`ParseUnit`
        the unit contained, as son nodes
    self.position : :obj:`Tuple` of (int, int)
        position of unit (line, col)
    self.value : :obj:`Any`
        reserved place for advanced usage
    self.unit_property : :obj:`Any`
        reserved place for advanced usage
    """

    def __init__(self, name: str, parse_units: List, position: Tuple[int, int], value: Any, unit_property: Any) -> None:
        """
        Unit of Parse, such as phrase.

        Parameters
        ----------
        name : str
            name of this, such as "class"
        parse_units : :obj:`list` of :obj:`ParseUnit`
            the unit contained, as son nodes
        position : :obj:`Tuple` of (int, int)
            position of unit (line, col)
        value : :obj:`Any`
            reserved place for advanced usage
        unit_property : :obj:`Any`
            reserved place for advanced usage
        """
        self.name = name
        self.parse_units = parse_units
        self.position = position
        self.value = value
        self.unit_property = unit_property

    def is_terminator(self) -> bool:
        """
        is terminator, if no children.
        """
        return len(self.parse_units) == 0


class Production:
    """
    the production of grammar

    Parameters
    ----------
    self.pre : str
        antecedent
    self.sufs : :obj:`list` of str
        seccedents
    self.semant_callback : :obj:`Callable` of :obj:`list` of :obj:`ParseUnit` with return of :obj:`Any`
        callback for semanting
    self.attr : :obj:`Any`
        attr for advansing usage
    """

    def __init__(self, pre: str, sufs: List[str], semant_callback: Callable[[List[ParseUnit]], Any], attr: Any) -> None:
        """
        the production of grammar

        Parameters
        ----------
        pre : str
            antecedent
        sufs : :obj:`list` of str
            seccedents
        semant_callback : :obj:`Callable` of :obj:`list` of :obj:`ParseUnit` with return of :obj:`Any`
            callback for semanting
        attr : :obj:`Any`
            attr for advansing usage
        """
        self.pre = pre
        self.sufs = sufs
        self.semant_callback = semant_callback
        self.attr = attr

    def deep_copy(self):
        """
        Deep copy self
        """
        return Production(self.pre, [i for i in self.sufs], self.semant_callback, self.attr)


class ParseIndexException(Exception):
    """
    Exception when index is not match
    """

    def __init__(self, index_p: int, index_l: int, *args: object) -> None:
        """
        Exception when index is not match

        Parameters
        ----------
        index_p : int
            index should be (of parser)
        index_l : int
            current index (of lexer_result)
        *args
            for super class Exception
        """
        super().__init__(*args)
        self.index_p = index_p
        self.index_l = index_l

    def __str__(self) -> str:
        return f"Index Error: Index should be {self.index_p} instead of {self.index_l}"


class WaitTimeoutException(Exception):
    """
    Exception when `parse_lex_unit_async` waits timeout
    """

    def __init__(self, index_p: int, index_l: int, *args: object) -> None:
        """
        Exception when `parse_lex_unit_async` waits timeout

        Parameters
        ----------
        index_p : int
            index should be (of parser)
        index_l : int
            current index (of lexer_result)
        *args
            for super class Exception
        """
        super().__init__(*args)
        self.index_p = index_p
        self.index_l = index_l

    def __str__(self) -> str:
        return f"Wait Timeout: Index processing: {self.index_p}, this index is {self.index_l}"


class ProductSentenceException(Exception):
    """
    Exception when `product_sentence` is illeagal.
    """

    def __init__(self, product_sentence: str, *args: object) -> None:
        """
        Exception when `product_sentence` is illeagal.

        Parameters
        ----------
        product_sentence : str
            Error Sentence
        *args
            for super class Exception
        """
        super().__init__(*args)
        self.product_sentence = product_sentence

    def __str__(self) -> str:
        return f"ProductSentence Error: sentence \"{self.product_sentence}\" is illeagal."


class ParserFramework:
    """
    The very base framework of parsing, provide very base methods for all kind of parsing.

    Parameters
    ----------
    self._index : int
        For the parsing order checking.
    self._stack : :obj:`LifoQueue`
        stack for PDA
    self.productions : :obj:`list` of :obj:`Production`
        collection of productions
    self.semant_callback_dict : :obj:`dict` of str to :obj:`Callable` of :obj:`list` of :obj:`ParseUnit` with return of :obj:`Any`
        Semant Callback Dictionary for string convert to callback.
    """

    def __init__(self) -> None:
        self._index = 0
        self._stack = LifoQueue()
        self.productions: List[Production] = []
        self.semant_callback_dict: Dict[str,
                                        Callable[[List[ParseUnit]], Any]] = {}

    def parse(self, parse_unit: ParseUnit) -> None:
        """
        parsing entrance abstract method, you need to override this method.

        Parameters
        ----------
        parse_unit : :obj:`ParseUnit`
            parse unit
        """
        pass

    ###################################
    # receive result of lexing region #
    ###################################

    def parse_lex_unit(self, lexer_result: LexerResult) -> None:
        """
        Receive LexerResult and Parse synchronously, order of `lexer_result.index` will be check.

        Parameters
        ----------
        lexer_result : :obj:`LexerResult`
            The Result of Lexer

        Raises
        ------
        ParseIndexException
            when index is not match
        """
        # check the order
        if lexer_result.index != self._index:
            raise ParseIndexException(self._index, lexer_result.index)
        # transfer to parse unit
        parse_unit = ParseUnit(lexer_result.name, [],
                               lexer_result.position, lexer_result.value, None)
        # to parse it
        self.parse(parse_unit)
        # move to next
        self._index += 1

    async def parse_lex_unit_async(self, lexer_result: LexerResult, timeout: int = 600) -> None:
        """
        Receive LexerResult and Parse asynchronously, order of `lexer_result.index` will be checked. The author suggests to use `asyncio.create_task()` to invoke this method/function. This impl is different from C# version, the inner includes a waiting loop for matching `lexer_result.index`.

        Parameters
        ----------
        lexer_result : :obj:`LexerResult`
            The Result of Lexer
        timeout : int
            The wait timeout seconds

        Raises
        ------
        WaitTimeoutException
            when this waits timeout for matching `lexer_result.index`.
        """
        # The author changed the impl of ansyc task instead of C# version
        # wait for index, this can be cause dead loop if you DO NOT correctly use this framework.
        timeout = 100 * timeout  # 10 minus default
        now = 0
        while lexer_result.index != self._index:
            time.sleep(0.01)
            now += 1
            if now > timeout:
                raise WaitTimeoutException(self._index, lexer_result.index)
        # now is this index time
        # transfer to parse unit
        parse_unit = ParseUnit(lexer_result.name, [],
                               lexer_result.position, lexer_result.value, None)
        # to parse it
        self.parse(parse_unit)
        # move to next
        self._index += 1

    ##############################
    # receive productions region #
    ##############################

    def add_production(self, production: Production) -> None:
        """
        the most direct way to add `production`

        Parameters
        ----------
        production : :obj:`Production`
            production
        """
        self.productions.append(production)

    def add_semant_callback_dict(self, name: str, semant_callback: Callable[[List[ParseUnit]], Any]) -> None:
        """
        part of indirect way of add `production`, add the `semant_callback` to dict.

        Parameters
        ----------
        name : str
            name of semant_callback, the key in dict
        semant_callback : :obj:`Callable` of :obj:`list` of :obj:`ParseUnit` with return :obj:`Any`
            the callback when reduce one production for semant analysis
        """
        self.semant_callback_dict[name] = semant_callback

    @staticmethod
    def none_semant_callback(parse_units: List[ParseUnit]) -> Any:
        return None

    def add_production_by_str(self, product_sentence: str):
        """
        indirect way of add production by string, format string `product_sentence` as:
        `Pre ->|:|| Suf0 Suf1 ... [@SemantCallbackName[$attr]]`

        `$attr` is as: `$key1=value1[&key2=value2[&...]]`

        like this:
        `E -> F G`

        or like this:
        `E -> E add E @SemantAdd`

        or like this:
        `E | E opt E @SementOpt`

        or like this:
        `E : Delimiter E Delimiter @Semant`

        or with attr in string (attr will be string type), just add `$attr` directly behand `@SemantDelegateName`:
        `E -> E F @Semant$priority=0&attr2=xxx`

        If you need Semant Callback, please add `semant_callback` use `add_semant_callback_dict` first.

        Parameters
        ----------
        product_sentence : str
            formated string

        Raises
        ------
        ProductSentenceException
            if `product_sentence` is illeagal
        """
        vs = product_sentence.split(" ")
        if vs[1] != "->" and vs[1] != "|" and vs[1] != ":":
            raise ProductSentenceException(product_sentence)
        last_index = len(vs) - 1
        semant_callback = self.none_semant_callback
        attr = {}
        if vs[last_index][0] == '@':
            s = vs[last_index][1:].split("$", 2)
            semant_callback = self.semant_callback_dict[s[0]]
            if len(s) == 2:
                if s[1]:
                    attr_s = s[1].split("&")
                    for attr_i in attr_s:
                        attr_i_s = attr_i.split("=")
                        if len(attr_i_s) > 1:
                            attr[attr_i_s[0]] = attr_i_s[1]
                        else:
                            attr[attr_i_s[0]] = None
            last_index -= 1
        sufs = vs[2:last_index + 1]
        production = Production(vs[0], sufs, semant_callback, attr)
        self.add_production(production)

    def add_production_by_multi_str(self, *product_sentences: str) -> None:
        """
        indirect way of add production by strings, each format string in `product_sentences` as:
        `Pre ->|:|| Suf0 Suf1 ... [@SemantCallbackName[$attr]]`

        like this:
        `E -> F G`

        or like this:
        `E -> E add E @SemantAdd`

        or like this:
        `E | E opt E @SementOpt`

        or like this:
        `E : Delimiter E Delimiter @Semant`

        or with attr in string (attr will be string type), just add `$attr` directly behand `@SemantDelegateName`:
        `E -> E F @Semant$deal`

        If you need Semant Callback, please add `semant_callback` use `add_semant_callback_dict` first.

        Parameters
        ----------
        product_sentences : :obj:`tuple` of str
            formated strings

        Raises
        ------
        ProductSentenceException
            if one product sentence in `product_sentences` is illeagal
        """
        for ps in product_sentences:
            self.add_production_by_str(ps)
